internal/thelma/app/autoupdate/installer/installer.go

```diff
@@ -0,0 +1,120 @@
+package installer
+
+import (
+	"fmt"
+	"github.com/broadinstitute/thelma/internal/thelma/app/autoupdate/releasebucket"
+	"github.com/broadinstitute/thelma/internal/thelma/app/autoupdate/releases"
+	"github.com/rs/zerolog/log"
+)
+
+// ResolvedVersions contains resolved version information for Thelma releases
+type ResolvedVersions struct {
+	// VersionAlias tag or version string that was resolved (eg. "latest", "1.2.3", "v1.2.3")
+	VersionAlias string
+	// TargetVersion the resolved canonical version of Thelma that should be installed (eg. "v1.2.3")
+	TargetVersion string
+	// CurrentVersion locally-installed version of Thelma; may be "" if this is
+	// a bootstrap/fresh install
+	CurrentVersion string
+}
+
+// UpdateNeeded return true if Thelma should be updated
+func (v ResolvedVersions) UpdateNeeded() bool {
+	return v.CurrentVersion != v.TargetVersion
+}
+
+type Installer interface {
+	// ResolveVersions expand a version or tag to a canonical Thelma version and identify
+	// the currently-installed Thelma version
+	ResolveVersions(versionOrTag string) (ResolvedVersions, error)
+	// UpdateThelma perform a Thelma update, obtaining a file lock first so multiple processes don't
+	// step on each other
+	UpdateThelma(versionOrTag string) error
+}
+
+func New(releasesDir releases.Dir, bucket releasebucket.ReleaseBucket) Installer {
+	return &installer{
+		dir:    releasesDir,
+		bucket: bucket,
+	}
+}
+
+type installer struct {
+	dir    releases.Dir
+	bucket releasebucket.ReleaseBucket
+}
+
+func (i *installer) ResolveVersions(versionOrTag string) (ResolvedVersions, error) {
+	targetVersion, err := i.bucket.ResolveTagOrVersion(versionOrTag)
+	if err != nil {
+		return ResolvedVersions{}, err
+	}
+
+	currentVersion, err := i.dir.CurrentVersion()
+	if err != nil {
+		log.Debug().Err(err).Msgf("Could not identify current installed version of Thelma; is this a fresh install?")
+	}
+
+	return ResolvedVersions{
+		VersionAlias:   versionOrTag,
+		CurrentVersion: currentVersion,
+		TargetVersion:  targetVersion,
+	}, nil
+}
+
+func (i *installer) UpdateThelma(versionOrTag string) error {
+	resolved, err := i.ResolveVersions(versionOrTag)
+	if err != nil {
+		return fmt.Errorf("error updating Thelma: %v", err)
+	}
+
+	if !resolved.UpdateNeeded() {
+		log.Info().Msgf("Thelma is already up-to-date (current version %q matches %q)", resolved.CurrentVersion, resolved.VersionAlias)
+		return nil
+	}
+
+	return i.dir.WithInstallerLock(func() error {
+		return i.updateThelmaUnsafe(versionOrTag)
+	})
+}
+
+// perform a Thelma update (without a lock)
+func (i *installer) updateThelmaUnsafe(versionOrTag string) error {
```

Comment:
just making sure i understand, this is called "unsafe" because it doesn't lock on its own, and then we're calling it safely by only calling it from *installer.UpdateThelma once we get a lock?

Reply:
yep 👍